samples/boards/stm32/bluetooth/interactive_gui/src/main.c

```diff
@@ -0,0 +1,473 @@
+
+static const struct device *const hci_uart_dev =
+	DEVICE_DT_GET(DT_CHOSEN(zephyr_bt_c2h_uart));
+static K_THREAD_STACK_DEFINE(tx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread tx_thread_data;
+static K_FIFO_DEFINE(tx_queue);
+
+/* RX in terms of bluetooth communication */
+static K_FIFO_DEFINE(uart_tx_queue);
+
+#define H4_CMD		0x01
+#define H4_ACL		0x02
+#define H4_SCO		0x03
+#define H4_EVT		0x04
+#define H4_ISO		0x05
+#define H4_ST_EXT_CMD	0x81
+#define H4_ST_VND_CMD	0xFF
+
+#define ST_IDLE		0 /* Waiting for packet type. */
+#define ST_HDR		1 /* Receiving packet header. */
+#define ST_PAYLOAD	2 /* Receiving packet payload. */
+#define ST_DISCARD	3 /* Dropping packet. */
+
+/* Length of a discard/flush buffer.
+ * This is sized to align with a BLE HCI packet:
+ * 1 byte H:4 header + 32 bytes ACL/event data
+ * Bigger values might overflow the stack since this is declared as a local
+ * variable, smaller ones will force the caller to call into discard more
+ * often.
+ */
+#define H4_DISCARD_LEN 33
+
+#define RESP_VENDOR_CODE_OFFSET	1
+#define RESP_LEN_OFFSET_LSB	2
+#define RESP_LEN_OFFSET_MSB	3
+#define RESP_CMDCODE_OFFSET	4
+#define RESP_STATUS_OFFSET	5
+#define RESP_PARAM_OFFSET	6
+
+/* Types of vendor codes */
+#define VENDOR_CODE_ERROR	0
+#define VENDOR_CODE_RESPONSE	1
+
+/* Commands */
+#define VENDOR_CMD_READ_VERSION		0x01
+#define VENDOR_CMD_BLUENRG_RESET	0x04
+#define VENDOR_CMD_HW_BOOTLOADER	0x05
+
+struct bt_hci_ext_cmd_hdr {
+	uint16_t opcode;
+	uint16_t param_len;
+} __packed;
+
+struct bt_vendor_cmd_hdr {
+	uint8_t opcode;
+	uint16_t param_len;
+} __packed;
+
+struct bt_vendor_rsp_hdr {
+	uint8_t vendor_code;
+	uint16_t param_len;
+	uint8_t opcode;
+	uint8_t status;
+	uint8_t params[2];
+} __packed;
+
+static int h4_send(struct net_buf *buf);
+
+static uint16_t parse_cmd(uint8_t *hci_buffer, uint16_t hci_pckt_len, uint8_t *buffer_out)
+{
+	uint16_t len = 0;
+	struct bt_vendor_cmd_hdr *hdr = (struct bt_vendor_cmd_hdr *) hci_buffer;
+	struct bt_vendor_rsp_hdr *rsp = (struct bt_vendor_rsp_hdr *) (buffer_out + 1);
+
+	buffer_out[0] = H4_ST_VND_CMD;
+	rsp->vendor_code = VENDOR_CODE_RESPONSE;
+	rsp->opcode = hdr->opcode;
+	rsp->status = 0;
+
+	switch (hdr->opcode) {
+	case VENDOR_CMD_READ_VERSION:
+		rsp->params[0] = KERNEL_VERSION_MAJOR;
+		if (KERNEL_PATCHLEVEL >= 9) {
+			rsp->params[1] = (KERNEL_VERSION_MINOR * 10) + 9;
+		} else {
+			rsp->params[1] = (KERNEL_VERSION_MINOR * 10) + KERNEL_PATCHLEVEL;
+		}
+		len = 2;
+		break;
+#if DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v1) || DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v2)
+	case VENDOR_CMD_BLUENRG_RESET:
+		bluenrg_bt_reset(0);
+		break;
+	case VENDOR_CMD_HW_BOOTLOADER:
+		bluenrg_bt_reset(1);
+		break;
+#endif /* DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v1) || DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v2) */
+	default:
+		rsp->vendor_code = VENDOR_CODE_ERROR;
+		rsp->status = BT_HCI_ERR_UNKNOWN_CMD;
+	}
+
+	len += 2; /* Status and Command code */
+	rsp->param_len = len;
```

Comment:
Missing `sys_cpu_to_le16(len)` here? (since `len` is `uint16_t`)

Reply:
Yes, thanks for notice.
Done.